In [2]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
     

from google.colab import files
uploaded = files.upload()

Saving games.csv to games.csv


Chess dataset

In [4]:
import io
df = pd.read_csv(io.BytesIO(uploaded['games.csv']),header=0, na_values='-')


In [7]:
df.describe(include="all")

id  rated    created_at  last_move_at         turns  \
count      20058  20058  2.005800e+04  2.005800e+04  20058.000000   
unique     19113      2           NaN           NaN           NaN   
top     XRuQPSzH   True           NaN           NaN           NaN   
freq           5  16155           NaN           NaN           NaN   
mean         NaN    NaN  1.483617e+12  1.483618e+12     60.465999   
std          NaN    NaN  2.850151e+10  2.850140e+10     33.570585   
min          NaN    NaN  1.376772e+12  1.376772e+12      1.000000   
25%          NaN    NaN  1.477548e+12  1.477548e+12     37.000000   
50%          NaN    NaN  1.496010e+12  1.496010e+12     55.000000   
75%          NaN    NaN  1.503170e+12  1.503170e+12     79.000000   
max          NaN    NaN  1.504493e+12  1.504494e+12    349.000000   

       victory_status winner increment_code white_id  white_rating black_id  \
count           20058  20058          20058    20058  20058.000000    20058   
unique              4      3            400     9438           NaN     9331   
top            resign  white           10+0  taranga           NaN  taranga   
freq            11147  10001           7721       72           NaN       82   
mean              NaN    NaN            NaN      NaN   1596.631868      NaN   
std               NaN    NaN            NaN      NaN    291.253376      NaN   
min               NaN    NaN            NaN      NaN    784.000000      NaN   
25%               NaN    NaN            NaN      NaN   1398.000000      NaN   
50%               NaN    NaN            NaN      NaN   1567.000000      NaN   
75%               NaN    NaN            NaN      NaN   1793.000000      NaN   
max               NaN    NaN            NaN      NaN   2700.000000      NaN   

        black_rating  moves opening_eco          opening_name   opening_ply  
count   20058.000000  20058       20058                 20058  20058.000000  
unique           NaN  18920         365                  1477           NaN  
top              NaN  e4 e5         A00  Van't Kruijs Opening           NaN  
freq             NaN     27        1007                   368           NaN  
mean     1588.831987    NaN         NaN                   NaN      4.816981  
std       291.036126    NaN         NaN                   NaN      2.797152  
min       789.000000    NaN         NaN                   NaN      1.000000  
25%      1391.000000    NaN         NaN                   NaN      3.000000  
50%      1562.000000    NaN         NaN                   NaN      4.000000  
75%      1784.000000    NaN         NaN                   NaN      6.000000  
max      2723.000000    NaN         NaN                   NaN     28.000000

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20058 entries, 0 to 20057
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              20058 non-null  object 
 1   rated           20058 non-null  bool   
 2   created_at      20058 non-null  float64
 3   last_move_at    20058 non-null  float64
 4   turns           20058 non-null  int64  
 5   victory_status  20058 non-null  object 
 6   winner          20058 non-null  object 
 7   increment_code  20058 non-null  object 
 8   white_id        20058 non-null  object 
 9   white_rating    20058 non-null  int64  
 10  black_id        20058 non-null  object 
 11  black_rating    20058 non-null  int64  
 12  moves           20058 non-null  object 
 13  opening_eco     20058 non-null  object 
 14  opening_name    20058 non-null  object 
 15  opening_ply     20058 non-null  int64  
dtypes: bool(1), float64(2), int64(4), object(9)
memory usage: 2.3+ MB


In [9]:
df['time']=df['last_move_at']-df['created_at']


In [10]:
k=0
for x in df.duplicated(keep='first'):
  if x:
    k=k+1
k

429

we will drop 429 duplicate rows which have occured more than once

In [11]:
df = df.drop_duplicates(keep='first')

we will now drop rows which dont have proper time recorded

In [12]:
df1 = df[df['time']>0]

In [13]:
df1.describe(include="all")

id  rated    created_at  last_move_at         turns  \
count      11096  11096  1.109600e+04  1.109600e+04  11096.000000   
unique     11056      2           NaN           NaN           NaN   
top     NQQpViEr   True           NaN           NaN           NaN   
freq           2   8957           NaN           NaN           NaN   
mean         NaN    NaN  1.483138e+12  1.483140e+12     61.923666   
std          NaN    NaN  3.253592e+10  3.253578e+10     34.216315   
min          NaN    NaN  1.376772e+12  1.376772e+12      1.000000   
25%          NaN    NaN  1.480153e+12  1.480154e+12     38.000000   
50%          NaN    NaN  1.498847e+12  1.498849e+12     57.000000   
75%          NaN    NaN  1.503599e+12  1.503599e+12     81.000000   
max          NaN    NaN  1.504493e+12  1.504494e+12    349.000000   

       victory_status winner increment_code white_id  white_rating black_id  \
count           11096  11096          11096    11096  11096.000000    11096   
unique              4      3            342     5690           NaN     5688   
top            resign  white           10+0     ssf7           NaN  docboss   
freq             6236   5551           4250       48           NaN       44   
mean              NaN    NaN            NaN      NaN   1615.380317      NaN   
std               NaN    NaN            NaN      NaN    288.354021      NaN   
min               NaN    NaN            NaN      NaN    784.000000      NaN   
25%               NaN    NaN            NaN      NaN   1426.000000      NaN   
50%               NaN    NaN            NaN      NaN   1589.000000      NaN   
75%               NaN    NaN            NaN      NaN   1804.000000      NaN   
max               NaN    NaN            NaN      NaN   2700.000000      NaN   

        black_rating  moves opening_eco          opening_name   opening_ply  \
count   11096.000000  11096       11096                 11096  11096.000000   
unique           NaN  10918         344                  1264           NaN   
top              NaN  e4 e5         A00  Van't Kruijs Opening           NaN   
freq             NaN     21         579                   238           NaN   
mean     1607.812004    NaN         NaN                   NaN      4.905912   
std       286.050814    NaN         NaN                   NaN      2.919249   
min       789.000000    NaN         NaN                   NaN      1.000000   
25%      1422.000000    NaN         NaN                   NaN      3.000000   
50%      1584.000000    NaN         NaN                   NaN      4.000000   
75%      1799.000000    NaN         NaN                   NaN      6.000000   
max      2621.000000    NaN         NaN                   NaN     28.000000   

                time  
count   1.109600e+04  
unique           NaN  
top              NaN  
freq             NaN  
mean    1.538472e+06  
std     6.388395e+06  
min     1.922000e+03  
25%     4.017578e+05  
50%     7.065730e+05  
75%     1.145629e+06  
max     6.058447e+08

In [14]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11096 entries, 14 to 20057
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              11096 non-null  object 
 1   rated           11096 non-null  bool   
 2   created_at      11096 non-null  float64
 3   last_move_at    11096 non-null  float64
 4   turns           11096 non-null  int64  
 5   victory_status  11096 non-null  object 
 6   winner          11096 non-null  object 
 7   increment_code  11096 non-null  object 
 8   white_id        11096 non-null  object 
 9   white_rating    11096 non-null  int64  
 10  black_id        11096 non-null  object 
 11  black_rating    11096 non-null  int64  
 12  moves           11096 non-null  object 
 13  opening_eco     11096 non-null  object 
 14  opening_name    11096 non-null  object 
 15  opening_ply     11096 non-null  int64  
 16  time            11096 non-null  float64
dtypes: bool(1), float64(3), int64(

No missing values

On chess servers (like Internet Chess Club and Free Internet Chess Server) the time control is usually expressed with a formula: time+increment. Where time is a fixed amount of time (in minutes) assigned at move 0 for the entire game, and increment (in seconds) is the amount added after each move.

When the increment is 0 the time control turns into a sudden death: so a 5+0 game is a game with sudden death at 5 minutes.

When the increment is bigger than 0 the time control is a Fischer delay: e.g. a 2+12 is a game with Fischer delay with 2 minutes + 12 seconds each move

In [15]:
incCode = np.asarray(df1['increment_code'])
turns = np.asarray(df1['turns'])
incCode

array(['15+15', '40+40', '40+40', ..., '10+0', '10+0', '10+0'],
      dtype=object)

In [16]:
incCode.size

11096

In [17]:
for i in range(0,incCode.size):
  x = incCode[i]
  k = x.index('+')
  n1 = float(x[0:k])
  n2 = float(x[k:])*turns[i]/60.0
  n = n1+n2
  incCode[i] = n


In [19]:
df1['increment_code'] = df1['increment_code'].astype("float64")
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11096 entries, 14 to 20057
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              11096 non-null  object 
 1   rated           11096 non-null  bool   
 2   created_at      11096 non-null  float64
 3   last_move_at    11096 non-null  float64
 4   turns           11096 non-null  int64  
 5   victory_status  11096 non-null  object 
 6   winner          11096 non-null  object 
 7   increment_code  11096 non-null  float64
 8   white_id        11096 non-null  object 
 9   white_rating    11096 non-null  int64  
 10  black_id        11096 non-null  object 
 11  black_rating    11096 non-null  int64  
 12  moves           11096 non-null  object 
 13  opening_eco     11096 non-null  object 
 14  opening_name    11096 non-null  object 
 15  opening_ply     11096 non-null  int64  
 16  time            11096 non-null  float64
dtypes: bool(1), float64(4), int64(

<ipython-input-19-3f8611546b82>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['increment_code'] = df1['increment_code'].astype("float64")


In [20]:
df1.describe(include="all")

id  rated    created_at  last_move_at         turns  \
count      11096  11096  1.109600e+04  1.109600e+04  11096.000000   
unique     11056      2           NaN           NaN           NaN   
top     NQQpViEr   True           NaN           NaN           NaN   
freq           2   8957           NaN           NaN           NaN   
mean         NaN    NaN  1.483138e+12  1.483140e+12     61.923666   
std          NaN    NaN  3.253592e+10  3.253578e+10     34.216315   
min          NaN    NaN  1.376772e+12  1.376772e+12      1.000000   
25%          NaN    NaN  1.480153e+12  1.480154e+12     38.000000   
50%          NaN    NaN  1.498847e+12  1.498849e+12     57.000000   
75%          NaN    NaN  1.503599e+12  1.503599e+12     81.000000   
max          NaN    NaN  1.504493e+12  1.504494e+12    349.000000   

       victory_status winner  increment_code white_id  white_rating black_id  \
count           11096  11096    11096.000000    11096  11096.000000    11096   
unique              4      3             NaN     5690           NaN     5688   
top            resign  white             NaN     ssf7           NaN  docboss   
freq             6236   5551             NaN       48           NaN       44   
mean              NaN    NaN       19.497407      NaN   1615.380317      NaN   
std               NaN    NaN       27.144375      NaN    288.354021      NaN   
min               NaN    NaN        0.533333      NaN    784.000000      NaN   
25%               NaN    NaN       10.000000      NaN   1426.000000      NaN   
50%               NaN    NaN       10.833333      NaN   1589.000000      NaN   
75%               NaN    NaN       20.000000      NaN   1804.000000      NaN   
max               NaN    NaN      591.000000      NaN   2700.000000      NaN   

        black_rating  moves opening_eco          opening_name   opening_ply  \
count   11096.000000  11096       11096                 11096  11096.000000   
unique           NaN  10918         344                  1264           NaN   
top              NaN  e4 e5         A00  Van't Kruijs Opening           NaN   
freq             NaN     21         579                   238           NaN   
mean     1607.812004    NaN         NaN                   NaN      4.905912   
std       286.050814    NaN         NaN                   NaN      2.919249   
min       789.000000    NaN         NaN                   NaN      1.000000   
25%      1422.000000    NaN         NaN                   NaN      3.000000   
50%      1584.000000    NaN         NaN                   NaN      4.000000   
75%      1799.000000    NaN         NaN                   NaN      6.000000   
max      2621.000000    NaN         NaN                   NaN     28.000000   

                time  
count   1.109600e+04  
unique           NaN  
top              NaN  
freq             NaN  
mean    1.538472e+06  
std     6.388395e+06  
min     1.922000e+03  
25%     4.017578e+05  
50%     7.065730e+05  
75%     1.145629e+06  
max     6.058447e+08

id created_at last_move_at white_id black_id moves opening_name won't matter

In [21]:
df2=df1.drop(df.columns[[0,2,3,8,10,12,14]],axis=1)

In [23]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11096 entries, 14 to 20057
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   rated           11096 non-null  bool   
 1   turns           11096 non-null  int64  
 2   victory_status  11096 non-null  object 
 3   winner          11096 non-null  object 
 4   increment_code  11096 non-null  float64
 5   white_rating    11096 non-null  int64  
 6   black_rating    11096 non-null  int64  
 7   opening_eco     11096 non-null  object 
 8   opening_ply     11096 non-null  int64  
 9   time            11096 non-null  float64
dtypes: bool(1), float64(2), int64(4), object(3)
memory usage: 877.7+ KB


Dealing with Non numeric Values

In [24]:
def handle_non_numerical_data(df):
    columns = df.columns.values
    for column in columns:
        text_digit_vals = {}
        def convert_to_int(val):
            return text_digit_vals[val]

        if df[column].dtype != np.int64 and df[column].dtype != np.float64:
            column_contents = df[column].values.tolist()
            unique_elements = set(column_contents)
            x = 0
            for unique in unique_elements:
                if unique not in text_digit_vals:
                    text_digit_vals[unique] = x
                    x+=1

            df[column] = list(map(convert_to_int, df[column]))

    return df

In [25]:
df3 = handle_non_numerical_data(df2)

In [26]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11096 entries, 14 to 20057
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   rated           11096 non-null  int64  
 1   turns           11096 non-null  int64  
 2   victory_status  11096 non-null  int64  
 3   winner          11096 non-null  int64  
 4   increment_code  11096 non-null  float64
 5   white_rating    11096 non-null  int64  
 6   black_rating    11096 non-null  int64  
 7   opening_eco     11096 non-null  int64  
 8   opening_ply     11096 non-null  int64  
 9   time            11096 non-null  float64
dtypes: float64(2), int64(8)
memory usage: 953.6 KB


standardization

In [36]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()  

df3 = scaler.fit_transform(df3)

In [43]:
df4 = pd.DataFrame(df3)

covmat

In [45]:
covmat = df4.cov() 
covmat

0         1         2         3         4         5         6  \
0  1.000090  0.083091 -0.014016  0.018479 -0.092841 -0.003941  0.028807   
1  0.083091  1.000090 -0.177914 -0.132857  0.039209  0.133703  0.165330   
2 -0.014016 -0.177914  1.000090 -0.004623  0.039220  0.142241  0.128223   
3  0.018479 -0.132857 -0.004623  1.000090 -0.027754  0.093192 -0.159114   
4 -0.092841  0.039209  0.039220 -0.027754  1.000090 -0.025302 -0.023869   
5 -0.003941  0.133703  0.142241  0.093192 -0.025302  1.000090  0.631465   
6  0.028807  0.165330  0.128223 -0.159114 -0.023869  0.631465  1.000090   
7 -0.005724  0.016429 -0.005923 -0.010132 -0.002706 -0.040930 -0.029159   
8 -0.023736  0.054193  0.087731  0.007584 -0.009466  0.291196  0.273130   
9 -0.039820  0.068186 -0.012738 -0.014203  0.157028 -0.003251 -0.006183   

          7         8         9  
0 -0.005724 -0.023736 -0.039820  
1  0.016429  0.054193  0.068186  
2 -0.005923  0.087731 -0.012738  
3 -0.010132  0.007584 -0.014203  
4 -0.002706 -0.009466  0.157028  
5 -0.040930  0.291196 -0.003251  
6 -0.029159  0.273130 -0.006183  
7  1.000090  0.100742  0.002273  
8  0.100742  1.000090  0.001763  
9  0.002273  0.001763  1.000090

In [47]:
type(covmat)

pandas.core.frame.DataFrame

eigen

In [48]:
covm = covmat.to_numpy()

In [49]:
from numpy.linalg import eig
eig(covm)

(array([1.9130778 , 0.31978831, 1.26531929, 1.20410462, 0.68407323,
        0.75418503, 0.83303247, 0.95223094, 1.02126764, 1.05382197]),
 array([[-2.23676296e-02, -4.25164530e-02, -6.01472230e-02,
         -4.95639394e-01,  1.90789356e-01,  1.74371921e-01,
         -1.20727836e-01,  7.91900183e-01, -1.77191642e-01,
          9.23677977e-02],
        [-2.02152612e-01,  2.95191743e-02, -6.27084494e-01,
         -2.21534642e-01, -6.65720081e-01, -1.60006078e-01,
         -1.71448743e-01, -4.68355454e-02, -1.23848075e-01,
         -1.54473855e-02],
        [-1.80705647e-01,  2.17420779e-02,  4.58109736e-01,
          3.23554388e-01, -5.40038174e-01, -1.68765752e-01,
          5.29232398e-03,  4.62182017e-01,  3.27256578e-01,
          1.07054080e-01],
        [ 7.16411347e-02,  2.59504841e-01,  4.17844991e-01,
          7.10353503e-03, -1.71700936e-01, -1.60617336e-01,
         -1.46236659e-01, -7.10176499e-02, -8.17932042e-01,
         -4.64677941e-02],
        [ 2.03174189e-02, -3.73430

turns and increment time have the least contribution to the information given to us with this data set because the eigen values corresponding to their rows is small compared to others
we will pca now to reduce the dimension by 2


In [50]:
from sklearn.decomposition import PCA



In [52]:
pca = PCA(8) 
projected = pca.fit_transform(df3)

In [53]:
projected

array([[-2.22063453,  0.08559417,  1.63513855, ..., -1.61348131,
         1.09058944,  1.19909239],
       [ 2.09338415, -0.53714715,  3.72267694, ..., -1.09836568,
        -0.62679086,  2.7110074 ],
       [-0.27462095,  0.24424136,  2.42143972, ..., -2.10785135,
        -0.18786862,  0.01846298],
       ...,
       [-2.258058  , -0.237319  , -0.82791354, ..., -0.18960534,
         0.0178141 ,  0.27842999],
       [-1.11155609,  0.01634318, -0.52659968, ...,  0.52393491,
        -0.22904975,  0.13163674],
       [-1.73374016,  1.29370903, -1.08362458, ..., -0.11258227,
         0.11260791,  0.29882373]])

In [55]:
sum(pca.explained_variance_ratio_)

0.8996228932906727

around 11 percent loss of information

plotting of 8 dimessions not possible